sunxi-toc0.c

```diff
@@ -10,36 +11,484 @@
+	}
+
+	ctx = EVP_PKEY_CTX_new(pkey, NULL);
+	if (!ctx) {
+		fprintf(stderr, "Failed to create EVP_PKEY_CTX\n");
+		goto cleanup;
+	}
+
+	if (EVP_PKEY_verify_init(ctx) <= 0 ||
+		EVP_PKEY_CTX_set_rsa_padding(ctx, RSA_PKCS1_PADDING) <= 0 ||
+		EVP_PKEY_CTX_set_signature_md(ctx, EVP_sha256()) <= 0) {
+		fprintf(stderr, "Failed to set up verification context\n");
+		goto cleanup;
+	}
+
+	if (EVP_PKEY_verify(ctx,
+						sigSequence->signature,
+						sizeof(sigSequence->signature),
+						cert_digest,
+						SHA256_DIGEST_LENGTH) <= 0) {
+		fprintf(stderr, "Bad certificate signature\n");
+		goto cleanup;
+	}
+
+	ret = 0;
+
+cleanup:
+	if (ctx) EVP_PKEY_CTX_free(ctx);
+	if (pkey) EVP_PKEY_free(pkey);
+	if (pkctx) EVP_PKEY_CTX_free(pkctx);
+	if (bld) OSSL_PARAM_BLD_free(bld);
+	if (pub_params) OSSL_PARAM_free(pub_params);
+	if (n) BN_free(n);
+	if (e) BN_free(e);
+	return ret;
+}
+
+
+static void parse_key_item(const uint8_t *buf, size_t len, FILE *outf) {
+	if (len < sizeof(struct toc0_key_item)) {
+		fprintf(stderr, "Error: Buffer too small (len=%zu)\n", len);
+		return;
+	}
+
+	const struct toc0_key_item *item = (const struct toc0_key_item *)buf;
+	size_t signed_len = item->sig - buf;
+	verify_signature(item->sig, item->sig_len, buf, signed_len, item->key0, item->key0_n_len, item->key0 + item->key0_n_len, item->key0_e_len, outf);
+}
+
+static void print_toc0_item(const toc0_item_info *item, const uint8_t *toc0_data, long file_size, const uint8_t *fw_data, size_t fw_size, FILE *outf) {
+	if (memcmp(item->end_marker, "IIE;", 4) != 0) {
+		fprintf(stderr, "Invalid TOC0 item end marker\n");
+		exit(EXIT_FAILURE);
+	}
+
+	if (item->offset + item->length > (uint32_t)file_size) {
+		fprintf(stderr, "Item offset and length exceed file size\n");
+		exit(EXIT_FAILURE);
+	}
+
+	fprintf(outf, "\nItem:\n");
+	fprintf(outf, "  Name: 0x%06x (%s)\n", item->name, item_name(item->name));
+	fprintf(outf, "  Offset: 0x%08x\n", item->offset);
+	fprintf(outf, "  Length: %u bytes\n", item->length);
+	fprintf(outf, "  Status: 0x%08x\n", item->status);
+	fprintf(outf, "  Type: %u\n", item->type);
+	fprintf(outf, "  Run Address: 0x%08x\n", item->load_addr);
+
+	const uint8_t *blob = toc0_data + item->offset;
+	switch (item->name) {
+	case TOC0_ITEM_NAME_CERT:
+		fprintf(outf, "  -> Certificate File\n");
+		parse_cert_for_rotpk(blob, item->length, outf);
+		parse_cert_item(blob, item->length, fw_data);
+		break;
+	case TOC0_ITEM_NAME_FW:
+		fprintf(outf, "  -> Signed Boot Firmware\n");
+		break;
+	case TOC0_ITEM_NAME_KEY:
+		fprintf(outf, "  -> Key Ladder or Key Item\n");
+		parse_key_item(blob, item->length, outf);
+		break;
+	default:
+		fprintf(stderr, "Unknown item type 0x%08x\n", item->name);
+		exit(EXIT_FAILURE);
+	}
+}
+
+int output_toc0_info(void *sector, FILE *inf, FILE *outf, bool verbose) {
+	fseek(inf, 0, SEEK_END);
+	long file_size = ftell(inf);
+	fseek(inf, 0, SEEK_SET);
+
+	if (file_size <= 0)
+		return -EINVAL;
+
+	uint8_t *toc0_data = malloc(file_size);
```

Comment:
Please gather all variable declarations at the beginning of the function. If you find this too far away, consider splitting up the function.

Reply:
Should be fixed now.